tla-types/src/main/scala/at/forsyte/apalache/tla/typecmp/BuilderUtil.scala

```diff
@@ -0,0 +1,46 @@
+package at.forsyte.apalache.tla.typecmp
+
+import at.forsyte.apalache.tla.lir.oper.TlaOper
+import at.forsyte.apalache.tla.lir.{NameEx, OperEx, TlaEx, Typed}
+import scalaz._
+import scalaz.Scalaz._
+
+/**
+ * Utility methods
+ *
+ * @author
+ *   Jure Kukovec
+ */
+object BuilderUtil {
+
+  // Constructs an OperEx expression for oper applied to args. Uses typeCmp to validate args and determine the result type
+  def buildInstruction(oper: TlaOper, typeCmp: typeComputation, args: TlaEx*): TlaEx = typeCmp(args) match {
+    case Left(err) => throw err
+    case Right(typeRes) =>
+      OperEx(oper, args: _*)(Typed(typeRes))
+  }
+
+  // Removes elem from the scope, as the scope only contains types of free variables
+  def markAsBound(elem: NameEx): InternalState[Unit] = State[MetaInfo, Unit] { mi: MetaInfo =>
+    (mi.copy(mi.nameScope - elem.name), ())
```

Comment:
Future-proofing. For a concrete example, I'm considering adding automatic coverage analysis (i.e. which "branches" assign to which variables) to the builder (very far off in the future), so that when you build an expression it is automatically correct w.r.t. assignments. In that case, I'd extend `MetaInfo` with e.g. another map, but `markAsBound` wouldn't have to care about other components of `MetaInfo`.

Reply:
👍